### PR TITLE
Hopefully last whack-a-unicode

### PR DESCRIPTION
**Pull request by lsetiawan**
## Overview

This PR continues on #188 and hopefully address #148. Now the test and actual application uses the same way of encoding/decoding the content from the database.

### CODE REVIEW

File: wof/WaterML_1_1.py [hunk 1]
```diff
@@ -4986,8 +4986,8 @@ def export(self, outfile, level, namespace_='', name_='siteCodeType', namespaced
         self.exportAttributes(outfile, level, already_processed, namespace_, name_='siteCodeType')
         if self.hasContent_():
             outfile.write(u'>')
-            outfile.write(u''.format(self.valueOf_))
-            self.exportChildren(outfile, level + 1, namespace_=u'', name_=u'siteCodeType', pretty_print=pretty_print)
+            outfile.write(u'{}'.format(self.valueOf_))
```
Comment (lsetiawan) [inline wof/WaterML_1_1.py]: @ocefpaf I had to fix this from your last commit since `siteCode` value was not showing up within the xml.

Comment (ocefpaf) [reply wof/WaterML_1_1.py]: We should always try to make these into tests if possible to avoid regressions.

Thanks for fixing it!

File: wof/apps/spyned_1_1.py [hunk 2]
```diff
@@ -209,7 +209,7 @@ def GetValuesForASiteObject(ctx, site, startDate=None, endDate=None, authToken=N
             try:
                 timeSeriesResponse = wof_inst.create_get_values_site_response(
                         site, startDate, endDate)
-                outStream = io.BytesIO()
+                outStream = io.StringIO()
```
Comment (ocefpaf) [inline wof/apps/spyned_1_1.py]: We should document that stream is expected to be `StringIO`! When dealing with streamed data I prefer `BytesIO` but I am 100% on Python 3 now so I don't have the problems we are seeing here :smile:

**Comment by lsetiawan** [comment]
Sigh... tests are failing with decode error now... I thought that it worked a month ago.. darn.. let's see.. hmm..

**Comment by ocefpaf** [comment]
> Sigh... tests are failing with decode error now... I thought that it worked a month ago.. darn.. let's see.. hmm..

The real issue here is that we did not find the "endpoints" of the data stream to do the conversion only at those points. (That happens b/c the code is quite convoluted and we are not the original authors. Although, at this point, not even the original authors could find the endpoints easily.)

We can only add more tests and try to remove as much of the "inner" conversions as we can.

PS: by conversion I mean the enconde/decode calls.

**Comment by lsetiawan** [comment]
I did it right, the test just didn't reflect a real life output. Ideally, we should use the database test that we now have, but for now, this is good. I need to focus on addressing the errors that @miguelcleon encounters. @ocefpaf please review this PR and merge when you think is good. Thanks 😄

File: wof/apps/spyned_1_1.py (hunk 2 above)
Comment (lsetiawan) [reply wof/apps/spyned_1_1.py]: I wish I can be 100% Python 3 😛, I wouldn't need to deal with the unicode craziness! lol

File: test/test_dao_1_1.py [hunk 3]
```diff
@@ -131,7 +131,7 @@ def __init__(self, **kwargs):
         UnitsID=96,
         UnitsName='degree celsius',
         UnitsType='Temperature',
-        UnitsAbbreviation='°C'),
+        UnitsAbbreviation=u'°C'),
```
Comment (ocefpaf) [inline test/test_dao_1_1.py]: If the goal is not not support Python <3.3, which I recommend, this is good. The `u''` is innocuous in `Python >=3.4`.

File: wof/WaterML_1_1.py [hunk 4]
```diff
@@ -405,7 +405,7 @@ def quote_xml(inStr):
 
 
 def quote_xml_aux(inStr):
-    inStr = inStr.decode('utf-8')
+    inStr = inStr.encode('utf-8').decode('utf-8')
```
Comment (ocefpaf) [inline wof/WaterML_1_1.py]: I am never really comfortable with this b/c, IMO, we should always treat incoming data as bytes, but there is no other way around this unless we start supporting Python 3.

TL;DR :+1:

File: wof/apps/spyned_1_1.py (hunk 2 above)
Comment (ocefpaf) [reply wof/apps/spyned_1_1.py]: The number of "non-ASCII compliant languages" in ODM2 is quite high to justify moving to Python 3, no? :wink: 

If language is not a strong argument think of the cute symbols and greek letters we can use :smile:

Comment (lsetiawan) [reply wof/apps/spyned_1_1.py]: > The number of "non-ASCII compliant languages" in ODM2 is quite high to justify moving to Python 3, no? 

It's all up to the bosses 😉  haha

> If language is not a strong argument think of the cute symbols and greek letters we can use 

Oh man! I would use more emojis in code in that case! 👍
